### DISCUSSION TITLE
How to assign syntax highlighting to specyfic file extension ?

### DISCUSSION THREAD

**Discussion by mlipok** [WinMerge/winmerge#542] [General]
Here:
https://github.com/WinMerge/winmerge/issues/446

I create FR but maybe there is a way to setup default highlighting by a file extension ?

**Comment by sdottaka** [comment]
Unfortunately, there is no way to do that.

**Reply by mlipok** [reply]
So which files are automaticaly opened with "Basic" higlightment ?

**Reply by sdottaka** [reply]
> So which files are automaticaly opened with "Basic" higlightment ?

Files with the extensions bas, vb, vbs, frm, ds, cls, ctl, pag, dsr are opened with Basic language syntax hightlightment.

`winmerge\Externals\crystaledit\editlib\parsers\crystallineparser.cpp`
~~~
	SRC_BASIC, _T ("Basic"), _T ("bas,vb,vbs,frm,dsm,cls,ctl,pag,dsr"), &ParseLineBasic, SRCOPT_AUTOINDENT, /*4,*/ _T (""), _T (""), _T ("\'"), (DWORD)-1,
~~~

**Reply by mlipok** [reply]
So is it possible to just modify them like this:

`SRC_BASIC, _T ("Basic"), _T ("bas,vb,vbs,frm,dsm,cls,ctl,pag,dsr,au3"), &ParseLineBasic, SRCOPT_AUTOINDENT, /*4,*/ _T (""), _T (""), _T ("\'"), (DWORD)-1,`

It would be just exactly what I need as a quick fix for my FR #446

**Reply by mlipok** [reply]
Pull request created:
https://github.com/WinMerge/winmerge/pull/543

**Reply by mlipok** [reply]
I have also longer solution:

Please duplicate this file content
`winmerge\Externals\crystaledit\editlib\parsers\basic.cpp`
to:
`winmerge\Externals\crystaledit\editlib\parsers\AutoIt3.cpp`

I will change them accordingly to AutoIt3 syntax.

**Reply by sdottaka** [reply]
I just added parsers/autoit.cpp with commit https://github.com/WinMerge/winmerge/commit/c5a124c23d027aad72c807093d9d330e730b642f

**Comment by tjmprm77** [comment]
I think it is useful to have a feature to set up default highlighting by a file extension in WinMerge.
Do you have any plans to implement that feature?
If you have no plans and there is no problem to add it, I will implement it and send PR.

**Reply by sdottaka** [reply]
I have no plans. Pull requests are welcome!

**Reply by tjmprm77** [reply]
Thank you for your reply.
I will start implementing it.

**Reply by datadiode** [reply]
You may want to model your implementation after https://github.com/datadiode/winmerge2011:

![winmerge2011_editor_lexers](https://user-images.githubusercontent.com/10423465/104747534-f38c7200-5750-11eb-8d2e-05b1f7464022.png)

I was going to create a PR for this myself once I'd find time, but nothing happened yet.

**Reply by tjmprm77** [reply]
Thank you for the very valuable information.
I didn't know that you had implemented the feature in WinMerge2011.
Although it cannot be ported as it is due to the difference in GUI component classes used in WinMerge and WinMerge 2011, I will implement it referring to your implementation below.
https://github.com/datadiode/winmerge2011/commit/17a2c0e85495b4cae7658446379ff54952c48a6c

**Reply by marceloalbertopiazza** [reply]
> I have no plans. Pull requests are welcome!

Hello!
How could I add support to syntax highlighting to another language (Clipper/xBase)?
I have the configuration file used by Notepad++ with this language and I think it could be useful to WinMerge.
Thanks!

**Reply by mlipok** [reply]
@sdottaka could add new "files", like he do recently in case of AutoIt, but you @marceloalbertopiazza will be responsible to update them.

**Reply by tjmprm77** [reply]
I sent a PR for a feature to set up default highlighting by a file extension yesterday and it was merged.

@datadiode
In implementing this feature, I had a hard time getting the list control's in-place combobox to work well. 
But by using the wildcard drop list you made, finally I was able to make it work properly.
I'm really thankful to you.

**Comment by Gitoffthelawn** [comment]
What do you think of exposing the string used in that line of code in the settings UI, so the user can modify the string containing the list of extensions?  Another tool I use does this, and it works quite well.

**Comment by Jayshua83** [comment]
So am I correct in assuming after reading this thread that WinMerge does NOT have Auto Syntax Highlighting? I have been using WinMerge2011 and I think it does? I would really like for WM to change the highlighting based on file type or whatever, having to tell it each and every time what I am looking at is ugh... lol

**Reply by datadiode** [reply]
> So am I correct in assuming after reading this thread that WinMerge does NOT have Auto Syntax Highlighting?

WinMerge applies syntax highlighting automatically if it recognizes the filename extension. The recognized filename extensions are customizable in WinMerge2011, and will probably soon be customizable in WinMerge too.

**Comment by mlipok** [comment]
get recent bulid:
https://ci.appveyor.com/project/sdottaka/winmerge/build/artifacts

which contains:
https://github.com/WinMerge/winmerge/pull/627

![image](https://user-images.githubusercontent.com/11089482/108128709-759feb80-70ad-11eb-8952-8a36c0ae7a33.png)
